File: 538_convert-bst-to-greater-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        int curSum = 0;
        convertBST(curSum, root);
        return root;
    }
    void convertBST(int& curSum, TreeNode* root) {
        if (!root) return;
        convertBST(curSum, root->right);
        curSum = (root->val += curSum);
        convertBST(curSum, root->left);
    }
};

File: 538_convert-bst-to-greater-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.cur = 0
        def dfs(root):
            if root:
                dfs(root.right)
                root.val += self.cur
                self.cur = root.val
                dfs(root.left)
        dfs(root)
        return root
